File: OLD Repos/Linear Algebra/Clustering/Clustering.jl
#### Clustering

using LinearAlgebra
using Statistics

function JCluster(X, reps, assignment)
     t = 0
     c = 0
     for i in 1:length(X)
          c = c + norm(X[i] - reps[assignment[i]])^2
          t = t + 1
     end
     r = c / t
     return r
end

function JCluster_plus(X, reps, assignment)
    return mean([norm(X[i] - reps[assignment[i]])^2 for i = 1:length(X)])
end

JCluster_pro(X, reps, assignment) = mean([norm(X[i] - reps[assignment[i]])^2 for i = 1:length(X)])

U = [[ 0, 1],
     [ 1, 0],
     [-1, 1]]

R = [[ 1, 1],
     [ 0, 0]]

A = [1, 2, 1]

re = JCluster(U, R, A)
println(re)             # 2.0

re = JCluster_plus(U, R, A)
println(re)             # 2.0

re = JCluster_pro(U, R, A)
println(re)             # 2.0

@time JCluster(U, R, A)
@time JCluster_plus(U, R, A)
@time JCluster_pro(U, R, A)

File: OLD Repos/Linear Algebra/Clustering/Clustering.py
#### Clustering

import math
import numpy as np

def norm(x):
    return np.linalg.norm(x)

def JCluster(X, reps, assignment):
    t = 0
    c = 0
    for i in range(0, len(X)):
        c = c + norm(X[i] - reps[assignment[i]])**2
        t = t + 1
    r = c / t
    return r

def JCluster_plus(X, reps, assignment):
    return np.mean([norm(X[i] - reps[assignment[i]])**2 for i in range(0, len(X))])

U = np.array([[ 0, 1],
              [ 1, 0],
              [-1, 1]])

R = np.array([[ 1, 1],
              [ 0, 0]])

S = np.array([1, 2 ,1])
A = S - 1

re = JCluster(U, R, A)
print(re)             # 2.0

re = JCluster_plus(U, R, A)
print(re)             # 2.0
